### PR TITLE
New interface for random.financial

### PR DESCRIPTION
**Pull request by tlsim**
* Implements the interface discussed in #763, allowing for controlling the time increment #754.
Time increment is provided as a number of seconds, not a d3 time interval - it's not possible to create a d3 time interval of 5 mins, for example.
* Replaces volume model with something simpler, and makes it configurable as a functor property.
* Component function now does not mutate state - stateful usage is provided by the closure returned from `generator.stream()`, as proposed.

Todo:
- [x] Convert usages
- [x] Improve documentation

### CODE REVIEW

**Comment by tlsim** [comment]
We have the use case of using this component to create an initial array, then add new data using the stream, and we want the open price of the first streamed point to be related to the last point of initial data.

What's preferred? 
```javascript
var generator = fc.data.random.financial()
  .start(new Date(2016, 0, 1));

// Currently
var data = generator(new Date(2016, 1, 1));
generator.startDate(new Date(2016, 1, 1));
generator.startPrice(/* something sensible */);
var stream = generator.stream();
data.push(stream());

// Or this?
var stream = generator.stream(); 
var data = stream(/* end date and/or number of points? */); // Returns array
data.push(stream()[0]);

// Or revert to having generator mutate date and price?
var data = generator(new Date(2016, 1, 1));
var stream = generator.stream();
data.push(stream());
```

**Comment by ColinEberhardt** [comment]
Thanks @tlsim - this is going to be a big improvement. Here's some notes on our discussion:

The generator shouldn't have an API that distinguishes between generating an array of items versus streaming single datapoints. One good way to solve this is to make everything a streaming operation.

```
// a factory that creates streams with a given initial state.
var generator = fc.data.random.financial()
  .start(new Date(2016, 0, 1));

// construct a stream from the generator - you can of course do this multiple times. 
var stream = generator();

// populate an array by asking for a number of points
var data = stream.next(20);

// stream new items into the array
data.push(stream.next());
data.push(stream.next());
data.push(stream.next());


// another way to populate the initial data
var stream2 = generator();
var data2 = stream2.until(function(datum) { datum.date > new Date(2015,12,12); });

```

**Comment by chrisprice** [comment]
One final (?!) tweak to the API -

```js
var dataGenerator = fc.data.random.financial();

var stream = dataGenerator.stream();

var snapshot = stream.take(2); // starts from start (first call to stream) [{}, {}]

var datum = stream.next(); // continues from above {}

stream.until(function(datum) { datum.date > new Date(2015,12,12); }) // continues from above [{},...]

var data = dataGenerator(2); // always starts from start date [{}, {}]
```

The above code tries to ensure consistent return types and allow the following 3 use cases -

* Generate n data sets of a fixed range
* Subscribe to a stream of data points
* Request a snapshot to back-fill a range, then stream in new data

**Comment by tlsim** [comment]
Updated.
One more thing to check:
`stream.until` takes function(datum) but `filter` takes function(date). This might feel inconsistent. Is this worth changing?

**Comment by ColinEberhardt** [comment]
> One more thing to check: stream.until takes function(datum) but filter takes function(date). This might feel inconsistent. Is this worth changing?

:+1: for changing the `filter` function for consistency.

**Comment by tlsim** [comment]
@ColinEberhardt @chrisprice thanks for the guidance! I think this is ready to review.

File: site/src/components/data/financial.md [hunk 1]
```diff
@@ -18,11 +23,28 @@ The following example:
...
 {{{financial-example-js}}}
-</script>
+</script> 
+
+{{{ dynamic-include 'codepen' html="financial-example-html" js="financial-example-js" }}}
+
+## Stream
+The component exposes a stateful streaming interface.
+You can see this in action in the [streaming chart example](../../examples/streaming/index.html).
+First create a stream from a random.financial instance with `financial.stream()`.
```
Comment (chrisprice) [inline site/src/components/data/financial.md]: I think this is a bit confusing, maybe reword it to -

First create a stream from an instance of ```fc.data.random.financial``` by calling the ```stream()``` method.

Comment (chrisprice) [inline site/src/components/data/financial.md]: Actually you've used the convention of methods without parenthesis below so stick to that (i.e. ```stream```).

File: site/src/components/data/financial.md [hunk 2]
```diff
@@ -18,11 +23,28 @@ The following example:
...
 {{{financial-example-html}}}
 <script type="text/javascript">
 {{{financial-example-js}}}
-</script>
+</script> 
+
+{{{ dynamic-include 'codepen' html="financial-example-html" js="financial-example-js" }}}
+
+## Stream
+The component exposes a stateful streaming interface.
```
Comment (chrisprice) [inline site/src/components/data/financial.md]: I would try and make it clearer what the difference between the stream and the base component is.

File: site/src/examples/streaming/streaming.js [hunk 3]
```diff
@@ -1,11 +1,10 @@
 // create some test data
```
Comment (chrisprice) [inline site/src/examples/streaming/streaming.js]: These are great examples of how much better things are with these changes!

File: src/data/random/financial.js [hunk 4]
```diff
@@ -1,155 +1,127 @@
...
-                volume: volumes[currentStep]
-            });
-            currentIntraStep += stepsPerDay;
-            currentStep += 1;
-            startDate.setUTCDate(startDate.getUTCDate() + 1);
-        }
-        return ohlcv;
+    var startDate = new Date();
+    var startPrice = 100;
+    var granularity = 60 * 60 * 24;
```
Comment (chrisprice) [inline src/data/random/financial.js]: This could be spun out as another issue but would it make sense to use [```d3.time.intervals```](https://github.com/mbostock/d3/wiki/Time-Intervals#interval) for this? Not only does it require mental gymnastics to parse the value, it's also in seconds rather than milli-seconds as everywhere else in JS.

File: src/data/random/financial.js [hunk 5]
```diff
@@ -1,155 +1,127 @@
...
-        return calculateOHLC(days, prices, volumes).filter(function(d) {
-            return filter(d.date);
-        });
+    function calculateOHLC(start, price) {
+        var prices = walk
+            .period(granularityInYears())(price);
+        var datum = {
+            date: start,
+            open: prices[0],
+            high: Math.max.apply({}, prices),
```
Comment (chrisprice) [inline src/data/random/financial.js]: I think the context argument should be ```Math```? E.g. this is what Babel produces for spread operators -
```js
Math.max(...[1,2,3])
/// --->
Math.max.apply(Math, [1, 2, 3]);
```

File: src/data/random/financial.js [hunk 6]
```diff
@@ -1,155 +1,127 @@
...
-    };
-    gen.startPrice = function(x) {
-        if (!arguments.length) {
-            return startPrice;
+    financial.stream = makeStream;
+
+    function makeStream() {
+        var latest;
+        function getDatum() {
+            if (!latest) {
```
Comment (chrisprice) [inline src/data/random/financial.js]: Another advantage of ```d3.time.interval``` would allow the offset to be stored, simplifying this a bit.

File: src/data/random/financial.js [hunk 7]
```diff
@@ -1,155 +1,127 @@
...
         }
-        startPrice = x;
-        return gen;
-    };
-    gen.startVolume = function(x) {
-        if (!arguments.length) {
-            return startVolume;
+        function getNextDatum() {
+            var ohlc = getDatum();
+            while (filter && !filter(ohlc)) {
```
Comment (chrisprice) [inline src/data/random/financial.js]: Would ```do/while``` simplify this?

File: src/data/random/financial.js [hunk 8]
```diff
@@ -1,155 +1,127 @@
...
+        };
+        stream.take = function(numPoints) {
+            var point;
+            var data = [];
+            for (point = 0; point < numPoints; point += 1) {
+                data.push(getNextDatum());
+            }
+            return data;
+        };
+        stream.until = function(comparison) {
```
Comment (chrisprice) [inline src/data/random/financial.js]: Would it make sense to pass ```d, i``` to the comparison function to match the d3 pattern? If so, could take just be a specialisation of this function?

File: src/data/random/financial.js [hunk 9]
```diff
@@ -1,155 +1,127 @@
...
+            var data = [];
+            if (!comparison || latest && comparison(latest)) {
+                return data;
+            }
+            var ohlc = getNextDatum();
+            while (comparison) {
```
Comment (chrisprice) [inline src/data/random/financial.js]: I can't quite decipher this from looking but this code looks like it's going to cause the value which fails the comparison test to be skipped on the next request for more data?

Comment (chrisprice) [inline src/data/random/financial.js]: Scratch that, I had misunderstood the API.

Comment (chrisprice) [inline src/data/random/financial.js]: Now I understand the API behaviour (stream.until should generate data up to and including the datum that satisfies the specified condition), I wonder if it should more closely match the [RxJS takeWhile behaviour](https://github.com/Reactive-Extensions/RxJS/blob/master/doc/api/core/operators/takewhile.md). Maybe another one for the issue pile?

File: src/data/random/financial.js (hunk 4 above)
Comment (tlsim) [inline src/data/random/financial.js]: Yeah I took another look and it seems possible. We would need to pass `step` to get multiples of each interval, and use `interval.offset(date, step)` internally.
E.g. 5 minute intervals:
```javascript
fc.data.random.financial()
  .startDate(new Date(2016, 0, 1))
  .startPrice(100)
  .interval([d3.time.minute, 5]);
```

File: src/data/random/financial.js (hunk 8 above)
Comment (tlsim) [inline src/data/random/financial.js]: NIce idea. I'll see what I can do.

File: src/data/random/financial.js (hunk 9 above)
Comment (tlsim) [inline src/data/random/financial.js]: Yeah I don't much like how this turned out - I think the takeUntil behavior would work, would just have to move setting the latest point out of `getDatum`, otherwise the next call to a stream method after an `until` call returns the date unexpectedly incremented.

**Comment by tlsim** [comment]
Updated to address comments. Matching RxJS takeUntil more closely has simplified things. Also now allows for setting the units of mu and sigma with an interval offset, so no more 'ms per year' constant.

**Comment by ColinEberhardt** [comment]
Thanks for the update @chrisprice are you ok to review?

**Comment by chrisprice** [comment]
LGTM :+1: 

Merging is going to break the site as per #873.
